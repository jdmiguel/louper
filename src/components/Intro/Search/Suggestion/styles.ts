import { styled } from '@mui/material/styles';
import Button from '@mui/material/Button';

export const StyledButton = styled(Button)(({ theme }) => ({
  '&:hover': {
    backgroundColor: theme.palette.primary.light,
  },
}));

export const StyledAvatarWrapper = styled('div')({
  height: 32,
  marginRight: 10,
  position: 'relative',
  width: 32,
});
