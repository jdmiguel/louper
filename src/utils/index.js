export const errorLiterals = {
  maximumRequest:
    'You have excedeed the maximum allowed request. Please, wait for a while',
  unavailableUser: 'Please, choose an available user',
};

export const externalLink = (url) => window.open(url, '_blank');
