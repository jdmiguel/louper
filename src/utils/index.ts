import { ResponseError } from './types';

enum ErrorMessage {
  MAX = 'You have excedeed the maximum allowed request. Please, wait for a while',
  NO_USER = 'Please, choose an available user',
  GENERIC = 'Sorry! there was an error on the server side.',
}

export const getErrorMessage = (error: ResponseError) => {
  if (error.status === 403) {
    return ErrorMessage.MAX;
  }

  if (error.status === 404) {
    return ErrorMessage.NO_USER;
  }

  return ErrorMessage.GENERIC;
};

export const debounce = (fn: any, delay: number) => {
  let timer: ReturnType<typeof setTimeout>;

  return (...args: any) => {
    if (timer) clearTimeout(timer);
    timer = setTimeout(() => {
      fn(...args);
    }, delay);
  };
};
