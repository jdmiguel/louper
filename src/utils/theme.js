/* material-ui */
import { keyframes } from '@emotion/react';
import { createTheme } from '@mui/material/styles';

export const theme = createTheme({
  overrides: {
    MuiCssBaseline: {
      '@global': {
        html: {
          overflowX: 'hidden',
          background: '#E0E6EF',
        },
        body: {
          margin: 0,
          padding: 0,
        },
        button: {
          boxShadow: 'initial !important',
        },
      },
    },
  },
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h1: {
      color: '#353D40',
      fontSize: '4.8rem',
      fontWeight: 700,
      letterSpacing: '-0.25rem',
      lineHeight: 0,
      '@media (min-width: 769px)': {
        fontSize: '6rem',
      },
    },
    h2: {
      color: '#353D40',
      fontSize: '3.5rem',
      fontWeight: 300,
      letterSpacing: '-0.1rem',
      lineHeight: 0,
      '@media (min-width: 769px)': {
        fontSize: '4.2rem',
      },
    },
    h3: {
      color: '#353D40',
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 769px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      color: '#353D40',
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 769px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      color: '#6b77a1',
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      color: '#757d81',
      fontSize: '0.8rem',
      fontWeight: 500,
      textTransform: 'uppercase',
    },
    button: {
      fontSize: '0.9rem',
    },
    body1: {
      color: '#353D40',
      fontSize: '0.9rem',
      lineHeight: '1.1rem',
    },
    body2: {
      color: '#E1D7EC',
      fontSize: '0.8rem',
      lineHeight: '1rem',
    },
    subtitle2: {
      color: '#353D40',
      fontSize: '0.8rem',
      fontWeight: 500,
      textTransform: 'uppercase',
    },
  },
  palette: {
    background: {
      default: '#E0E6EF',
    },
    primary: {
      main: '#553285',
      light: '#7B52AB',
      dark: '#36175E',
      contrastText: '#E1D7EC',
    },
    secondary: {
      main: '#6b77a1',
      light: '#c7cde1',
      dark: '#565e7a',
      contrastText: '#bac1d9',
    },
    neutral: {
      main: '#353D40',
    },
  },
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
  },
});
