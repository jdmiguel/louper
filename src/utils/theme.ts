/* material-ui */
import { keyframes } from '@emotion/react';
import { createTheme } from '@mui/material/styles';

/* colors */
import { colors } from './colors';

type Animation = {
  greet: any;
  shimmer: any;
};

declare module '@mui/material/styles' {
  interface Theme {
    animation: Animation;
    overrides: any;
  }
  interface ThemeOptions {
    animation: Animation;
    overrides: any;
  }
  interface Palette {
    neutral: Palette['primary'];
    overlay: Palette['primary'];
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary'];
    overlay: PaletteOptions['primary'];
  }
}

export const theme = createTheme({
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
    shimmer: keyframes`
      0% { background-position-x: -1200px}
      100% { background-position-x: 1200px}
    `,
  },
  overrides: [
    {
      MuiCssBaseline: {
        '@global': {
          html: {
            overflowX: 'hidden',
            background: colors.darkBlue,
          },
          body: {
            margin: 0,
            padding: 0,
          },
          button: {
            boxShadow: 'initial !important',
          },
        },
      },
    },
  ],
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h2: {
      fontSize: '1.1rem',
      fontWeight: 300,
      '@media (min-width: 768px)': {
        fontSize: '1.3rem',
      },
    },
    h3: {
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 768px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 768px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      fontSize: '1rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    body1: {
      fontSize: '1em',
    },
    body2: {
      fontSize: '0.9em',
    },
    overline: {
      fontSize: '0.8rem',
      fontWeight: 700,
      lineHeight: '1rem',
      textTransform: 'uppercase',
    },
  },
  palette: {
    primary: {
      main: colors.pink,
      light: colors.lightPink,
      dark: colors.darkPink,
    },
    secondary: {
      main: colors.blue,
      light: colors.lightBlue,
      dark: colors.darkBlue,
    },
    neutral: {
      main: colors.grey,
      light: colors.lightGrey,
    },
    overlay: {
      main: colors.darkPinkOverlay,
    },
    background: {
      default: colors.darkBlue,
    },
    text: {
      primary: colors.white,
      secondary: colors.lightBlue,
    },
  },
});
