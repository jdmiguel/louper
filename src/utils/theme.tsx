import { ReactElement } from 'react';
import { keyframes } from '@emotion/react';
import { Keyframes } from '@emotion/serialize';
import { createTheme, ThemeProvider } from '@mui/material/styles';
import { colors } from './colors';

type Animation = {
  greet: Keyframes;
  fadeIn: Keyframes;
  fadeInUp: Keyframes;
  fadeInLeft: Keyframes;
  fadeInRight: Keyframes;
};

declare module '@mui/material/styles' {
  interface Theme {
    animation: Animation;
  }
  interface ThemeOptions {
    animation: Animation;
  }
  interface Palette {
    neutral: Palette['primary'];
    overlay: Palette['primary'];
  }
  interface PaletteOptions {
    neutral: PaletteOptions['primary'];
    overlay: PaletteOptions['primary'];
  }
}

export const theme = createTheme({
  animation: {
    greet: keyframes`
      0%,100% { transform: rotate(0) }
      20%,60% { transform: rotate(-25deg) }
      40%,80% { transform: rotate(10deg) }
    `,
    fadeIn: keyframes`
      0% { opacity:0 }
      100% { opacity:1 }
    `,
    fadeInUp: keyframes`
      0% { opacity:0; transform: translateY(20px); }
      100% { opacity:1; transform: translateY(0px);  }
    `,
    fadeInLeft: keyframes`
      0% { opacity:0; transform: translateX(-50px); }
      100% { opacity:1; transform: translateX(0px);  }
    `,
    fadeInRight: keyframes`
      0% { opacity:0; transform: translateX(50px); }
      100% { opacity:1; transform: translateX(0px);  }
    `,
  },
  components: {
    MuiCssBaseline: {
      styleOverrides: {
        body: {
          overflowX: 'hidden',
        },
        h1: {
          margin: 0,
          lineHeight: 0,
        },
        h2: {
          margin: 0,
          lineHeight: 0,
        },
      },
    },
  },
  typography: {
    fontFamily: ['Comfortaa', 'Arial', 'sans-serif'].join(','),
    htmlFontSize: 12,
    h2: {
      fontSize: '0.9rem',
      '@media (min-width: 480px)': {
        fontSize: '1rem',
      },
      '@media (min-width: 768px)': {
        fontSize: '1.1rem',
      },
      '@media (min-width: 1440px)': {
        fontSize: '1.15rem',
      },
      fontWeight: 300,
    },
    h3: {
      fontSize: '1.3rem',
      fontWeight: 700,
      '@media (min-width: 768px)': {
        fontSize: '1.5rem',
      },
    },
    h4: {
      fontSize: '1.1rem',
      fontWeight: 500,
      '@media (min-width: 768px)': {
        fontSize: '1.2rem',
      },
    },
    h5: {
      fontSize: '1rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    h6: {
      fontSize: '0.9rem',
      fontWeight: 700,
      textTransform: 'uppercase',
    },
    body1: {
      fontSize: '1rem',
    },
    body2: {
      fontSize: '0.9rem',
    },
    overline: {
      fontSize: '0.8rem',
      fontWeight: 700,
      lineHeight: '1rem',
      textTransform: 'uppercase',
    },
  },
  palette: {
    primary: {
      main: colors.pink,
      light: colors.lightPink,
      dark: colors.darkPink,
    },
    secondary: {
      main: colors.blue,
      light: colors.lightBlue,
      dark: colors.darkBlue,
    },
    error: {
      main: colors.red,
    },
    neutral: {
      dark: colors.grey,
      main: colors.lightGrey,
      light: colors.white,
    },
    overlay: {
      main: colors.darkPinkOverlay,
    },
    background: {
      default: colors.darkBlue,
      paper: colors.gradient,
    },
    text: {
      primary: colors.white,
      secondary: colors.lightBlue,
    },
  },
});

export const renderWithTheme = (component: ReactElement) => (
  <ThemeProvider theme={theme}>{component}</ThemeProvider>
);
