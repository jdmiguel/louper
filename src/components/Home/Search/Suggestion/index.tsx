import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import { BasicUser } from '@/utils/types';
import { StyledButton, StyledAvatarWrapper } from './styles';

type Props = {
  data: BasicUser;
  onClick: (userName: string) => void;
};

const Suggestion = ({ data, onClick }: Props) => (
  <StyledButton variant="contained" onClick={() => onClick(data.login)}>
    <Stack direction="row" sx={{ alignItems: 'center', width: '100%' }}>
      <StyledAvatarWrapper>
        <Avatar variant="circular" sx={{ height: 32, width: 32, position: 'absolute' }} />
        <Avatar
          alt="user avatar"
          variant="circular"
          src={data.avatar_url}
          sx={{ height: 32, width: 32, position: 'absolute' }}
        />
      </StyledAvatarWrapper>
      <Typography
        variant="h5"
        sx={{
          maxWidth: 100,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap',
        }}
      >
        {data.login}
      </Typography>
    </Stack>
  </StyledButton>
);

export default Suggestion;
