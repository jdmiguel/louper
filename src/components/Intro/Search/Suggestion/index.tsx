import { styled } from '@mui/material/styles';
import Stack from '@mui/material/Stack';
import Typography from '@mui/material/Typography';
import Avatar from '@mui/material/Avatar';
import Button from '@mui/material/Button';
import { User } from '@/utils/types';

const StyledButton = styled(Button)(({ theme }) => ({
  height: 44,
  minWidth: 160,
  '&:hover': {
    backgroundColor: theme.palette.primary.light,
  },
  '@media (min-width: 768px)': {
    minWidth: 180,
  },
}));

const AvatarWrapper = styled('div')({
  height: 32,
  marginRight: 10,
  position: 'relative',
  width: 32,
});

type Props = {
  data: User;
  onClick: (userName: string) => void;
};

const Suggestion = ({ data, onClick }: Props) => (
  <StyledButton variant="contained" onClick={() => onClick(data.login)}>
    <Stack direction="row" sx={{ alignItems: 'center', width: '100%' }}>
      <AvatarWrapper>
        <Avatar variant="circular" sx={{ height: 32, width: 32, position: 'absolute' }} />
        <Avatar
          alt="user avatar"
          variant="circular"
          src={data.avatar_url}
          sx={{ height: 32, width: 32, position: 'absolute' }}
        />
      </AvatarWrapper>
      <Typography
        variant="h5"
        sx={{
          maxWidth: 100,
          overflow: 'hidden',
          textOverflow: 'ellipsis',
          whiteSpace: 'nowrap',
        }}
      >
        {data.login}
      </Typography>
    </Stack>
  </StyledButton>
);

export default Suggestion;
