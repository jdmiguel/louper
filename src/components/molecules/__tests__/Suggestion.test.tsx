import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { renderWithTheme } from '../../../utils/theme';
import userEvent from '@testing-library/user-event';
import Suggestion from '../Suggestion';

describe('<Suggestion />', () => {
  const props = {
    data: {
      id: 1,
      login: 'John Doe',
      html_url: 'https://github.com/JohnDoe',
      avatar_url: 'https://github.com/JohnDoe.jpg',
    },
    onClick: jest.fn(),
  };

  it('renders the content properly', () => {
    const { container } = render(renderWithTheme(<Suggestion {...props} />));
    expect(container).toMatchSnapshot();
  });

  it('calls the correct callback when the enter key is pressed', async () => {
    render(<Suggestion {...props} />);

    const button = screen.getByRole('button');
    await userEvent.click(button);

    expect(props.onClick).toHaveBeenCalledWith('John Doe');
  });
});
