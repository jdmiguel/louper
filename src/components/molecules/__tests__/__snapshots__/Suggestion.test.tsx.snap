// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Suggestion /> renders the content properly 1`] = `
<div>
  <button
    class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root css-19efffc-MuiButtonBase-root-MuiButton-root"
    tabindex="0"
    type="button"
  >
    <div
      class="css-u0quxs-MuiStack-root"
    >
      <div
        class="css-195cbcv"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1d4sk8a-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-4p5aps-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <div
          class="MuiAvatar-root MuiAvatar-circular css-15vrgjq-MuiAvatar-root"
        >
          <img
            alt="user following avatar"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://github.com/JohnDoe.jpg"
          />
        </div>
      </div>
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1ntgs97-MuiTypography-root"
      >
        John Doe
      </h5>
    </div>
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </button>
</div>
`;
